crates/polars-ops/src/frame/pivot/mod.rs

```diff
@@ -187,117 +187,155 @@ fn pivot_impl(
     // used as separator/delimiter in generated column names.
     separator: Option<&str>,
 ) -> PolarsResult<DataFrame> {
-    let sep = separator.unwrap_or("_");
     polars_ensure!(!index.is_empty(), ComputeError: "index cannot be zero length");
+    polars_ensure!(!columns.is_empty(), ComputeError: "columns cannot be zero length");
+    if !stable {
+        println!("unstable pivot not yet supported, using stable pivot");
+    };
+    if columns.len() > 1 {
+        let schema = Arc::new(pivot_df.schema());
+        let binding = pivot_df.select_with_schema(columns, &schema)?;
+        let fields = binding.get_columns();
+        let column = format!("{{\"{}\"}}", columns.join("\",\""));
+        if schema.contains(column.as_str()) {
+            polars_bail!(ComputeError: "cannot use column name {column} that \
+            already exists in the DataFrame. Please rename it prior to calling `pivot`.")
+        }
+        let columns_struct = StructChunked::new(&column, fields).unwrap().into_series();
+        let mut binding = pivot_df.clone();
+        let pivot_df = unsafe { binding.with_column_unchecked(columns_struct) };
+        pivot_impl_single_column(
+            pivot_df,
+            &column,
+            values,
+            index,
+            agg_fn,
+            sort_columns,
+            separator,
+        )
+    } else {
+        pivot_impl_single_column(
+            pivot_df,
+            unsafe { columns.get_unchecked(0) },
+            values,
+            index,
+            agg_fn,
+            sort_columns,
+            separator,
+        )
+    }
+}
 
+fn pivot_impl_single_column(
+    pivot_df: &DataFrame,
+    column: &str,
+    values: &[String],
+    index: &[String],
+    agg_fn: Option<PivotAgg>,
+    sort_columns: bool,
+    separator: Option<&str>,
+) -> PolarsResult<DataFrame> {
+    let sep = separator.unwrap_or("_");
     let mut final_cols = vec![];
-
     let mut count = 0;
     let out: PolarsResult<()> = POOL.install(|| {
-        for column_column_name in columns {
-            let mut group_by = index.to_vec();
-            group_by.push(column_column_name.clone());
+        let mut group_by = index.to_vec();
+        group_by.push(column.to_string());
 
-            let groups = pivot_df.group_by_stable(group_by)?.take_groups();
+        let groups = pivot_df.group_by_stable(group_by)?.take_groups();
 
-            // these are the row locations
-            if !stable {
-                println!("unstable pivot not yet supported, using stable pivot");
-            };
```

Comment:
this comment comes from https://github.com/pola-rs/polars/pull/3199/files, but I think it's no longer relevant as there's now a variable called `row_locations` as it's clear what it is